grails-app/services/org/pih/warehouse/shipping/ShipmentService.groovy

```diff
@@ -704,17 +704,14 @@ class ShipmentService {
             log.info("Checking shipment item ${shipmentItem?.inventoryItem} quantity [" +
                     shipmentItem.quantity + "] <= quantity on hand [" + quantityOnHand + "]")
             if (duplicatedShipmentItemsQuantity > quantityOnHand && origin.supports(ActivityCode.MANAGE_INVENTORY)) {
-                shipmentItem.errors.rejectValue("quantity", "shipmentItem.quantity.cannotExceedAvailableQuantity",
-                        [
-                                shipmentItem.quantity + " " + shipmentItem?.product?.unitOfMeasure,
-                                quantityOnHand + " " + shipmentItem?.product?.unitOfMeasure,
-                                shipmentItem?.product?.productCode,
-                                shipmentItem?.inventoryItem?.lotNumber,
-                                origin.name,
-                                shipmentItem?.binLocation?.name ?: 'Default'
-                        ].toArray(),
-                        "Shipping quantity cannot exceed on-hand quantity for product code " + shipmentItem.product.productCode + " " + allPendingShipmentsWithProduct.shipment.shipmentNumber.unique())
-                throw new ValidationException("Shipment item is invalid", shipmentItem.errors)
+                String errorMessage = "Shipping quantity can not exceed on hand quantity for the following lines: " +
+                        "Product Code ${shipmentItem.product.productCode} " +
+                        "Lot ${shipmentItem?.inventoryItem?.lotNumber} " +
+                        "Bin ${shipmentItem?.binLocation?.name ?: 'Default'} " +
+                        "Qty ${shipmentItem?.quantity}. " +
+                        "This can occur if changes were made to inventory after this shipment was picked but before it shipped. " +
+                        "To move forward, please remove the lines above from the shipment or reduce to reflect current QOH."
+                throw new Exception(errorMessage)
```

Comment:
First of all ValidationException is a runtime exception. So switching to a generic exception might cause weird behavior. For example, runtime exceptions cause the current transaction to rollback. 

Secondly, I was hoping we could do this in a less hacky way. Like rendering the errors object so that we can take advantage of the localization of the error messages, but do so in a less verbose manner.

It actually looks like Grails has a Json renderer for this (not sure if it's available in Grails 1.3.9). Either way, it would be great to get this to work in a generic way.

https://github.com/grails/grails-core/blob/master/grails-plugin-rest/src/main/groovy/grails/rest/render/errors/VndErrorJsonRenderer.groovy

https://github.com/blongden/vnd.error

Reply:
@DominikaHatala Here's more documentation explaining what we should be doing here. This is what I'd like to see for validation exceptions. http://docs.grails.org/3.3.10/guide/REST.html#vndError

On validation errors (or any errors for that matter), the API should return a response like the following and the React app should be able to deal with it appropriately.
```{
  "errors": [
    {
      "object": "rest.test.Book",
      "field": "title",
      "rejected-value": null,
      "message": "Property [title] of class [class rest.test.Book] cannot be null"
    }
  ]
}
```

If we can't find an easy way to do that with Grails 1.3.9 then let me know and I'll create a ticket in OGBM to make it work there.

Reply:
Would it be possible to add a new JsonValidationException which will extend ValidationException and then handle it in a separate action in the ErrorsController?

Reply:
@awalkowiak I don't think that's even necessary. I think we can do this with the ValidationException. I just haven't really put any thought into to figure out how to get it to work. I'll spend a few minutes now and see if I can get it to work.

Reply:
So the response for validation errors actually had everything we needed. We just needed to put the pieces together. I'm now returning an errorMessages attribute in the JSON that should include the resolved validation error. I'd 
```
"errorMessages": [
    "Property [category] of class [Product] cannot be null",
    "Property [Name] of class [Product] cannot be null"
  ]
```

Here's the full JSON

```
$ curl -b cookies.txt -X GET -H "Accept: application/json" -H "Content-Type: application/json" https://openboxes.ngrok.io/openboxes/api/products/10001 | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  5454    0  5454    0     0  33875      0 --:--:-- --:--:-- --:--:-- 33875
{
  "errorCode": 400,
  "errorMessage": "Validation error. invalid product:\n- Field error in object 'org.pih.warehouse.product.Product' on field 'category': rejected value [null]; codes [org.pih.warehouse.product.Product.category.nullable.error.org.pih.warehouse.product.Product.category,org.pih.warehouse.product.Product.category.nullable.error.category,org.pih.warehouse.product.Product.category.nullable.error.org.pih.warehouse.product.Category,org.pih.warehouse.product.Product.category.nullable.error,product.category.nullable.error.org.pih.warehouse.product.Product.category,product.category.nullable.error.category,product.category.nullable.error.org.pih.warehouse.product.Category,product.category.nullable.error,org.pih.warehouse.product.Product.category.nullable.org.pih.warehouse.product.Product.category,org.pih.warehouse.product.Product.category.nullable.category,org.pih.warehouse.product.Product.category.nullable.org.pih.warehouse.product.Category,org.pih.warehouse.product.Product.category.nullable,product.category.nullable.org.pih.warehouse.product.Product.category,product.category.nullable.category,product.category.nullable.org.pih.warehouse.product.Category,product.category.nullable,nullable.org.pih.warehouse.product.Product.category,nullable.category,nullable.org.pih.warehouse.product.Category,nullable]; arguments [category,Product]; default message [Property [{0}] of class [{1}] cannot be null]\n- Field error in object 'org.pih.warehouse.product.Product' on field 'name': rejected value [null]; codes [org.pih.warehouse.product.Product.name.nullable.error.org.pih.warehouse.product.Product.name,org.pih.warehouse.product.Product.name.nullable.error.name,org.pih.warehouse.product.Product.name.nullable.error.java.lang.String,org.pih.warehouse.product.Product.name.nullable.error,product.name.nullable.error.org.pih.warehouse.product.Product.name,product.name.nullable.error.name,product.name.nullable.error.java.lang.String,product.name.nullable.error,org.pih.warehouse.product.Product.name.nullable.org.pih.warehouse.product.Product.name,org.pih.warehouse.product.Product.name.nullable.name,org.pih.warehouse.product.Product.name.nullable.java.lang.String,org.pih.warehouse.product.Product.name.nullable,product.name.nullable.org.pih.warehouse.product.Product.name,product.name.nullable.name,product.name.nullable.java.lang.String,product.name.nullable,nullable.org.pih.warehouse.product.Product.name,nullable.name,nullable.java.lang.String,nullable]; arguments [Name,Product]; default message [Property [{0}] of class [{1}] cannot be null]\n",
  "data": [
    {
      "arguments": [
        "category",
        "Product"
      ],
      "bindingFailure": false,
      "class": "org.springframework.validation.FieldError",
      "code": "nullable",
      "codes": [
        "org.pih.warehouse.product.Product.category.nullable.error.org.pih.warehouse.product.Product.category",
        "org.pih.warehouse.product.Product.category.nullable.error.category",
        "org.pih.warehouse.product.Product.category.nullable.error.org.pih.warehouse.product.Category",
        "org.pih.warehouse.product.Product.category.nullable.error",
        "product.category.nullable.error.org.pih.warehouse.product.Product.category",
        "product.category.nullable.error.category",
        "product.category.nullable.error.org.pih.warehouse.product.Category",
        "product.category.nullable.error",
        "org.pih.warehouse.product.Product.category.nullable.org.pih.warehouse.product.Product.category",
        "org.pih.warehouse.product.Product.category.nullable.category",
        "org.pih.warehouse.product.Product.category.nullable.org.pih.warehouse.product.Category",
        "org.pih.warehouse.product.Product.category.nullable",
        "product.category.nullable.org.pih.warehouse.product.Product.category",
        "product.category.nullable.category",
        "product.category.nullable.org.pih.warehouse.product.Category",
        "product.category.nullable",
        "nullable.org.pih.warehouse.product.Product.category",
        "nullable.category",
        "nullable.org.pih.warehouse.product.Category",
        "nullable"
      ],
      "defaultMessage": "Property [{0}] of class [{1}] cannot be null",
      "field": "category",
      "objectName": "org.pih.warehouse.product.Product",
      "rejectedValue": null
    },
    {
      "arguments": [
        "Name",
        "Product"
      ],
      "bindingFailure": false,
      "class": "org.springframework.validation.FieldError",
      "code": "nullable",
      "codes": [
        "org.pih.warehouse.product.Product.name.nullable.error.org.pih.warehouse.product.Product.name",
        "org.pih.warehouse.product.Product.name.nullable.error.name",
        "org.pih.warehouse.product.Product.name.nullable.error.java.lang.String",
        "org.pih.warehouse.product.Product.name.nullable.error",
        "product.name.nullable.error.org.pih.warehouse.product.Product.name",
        "product.name.nullable.error.name",
        "product.name.nullable.error.java.lang.String",
        "product.name.nullable.error",
        "org.pih.warehouse.product.Product.name.nullable.org.pih.warehouse.product.Product.name",
        "org.pih.warehouse.product.Product.name.nullable.name",
        "org.pih.warehouse.product.Product.name.nullable.java.lang.String",
        "org.pih.warehouse.product.Product.name.nullable",
        "product.name.nullable.org.pih.warehouse.product.Product.name",
        "product.name.nullable.name",
        "product.name.nullable.java.lang.String",
        "product.name.nullable",
        "nullable.org.pih.warehouse.product.Product.name",
        "nullable.name",
        "nullable.java.lang.String",
        "nullable"
      ],
      "defaultMessage": "Property [{0}] of class [{1}] cannot be null",
      "field": "name",
      "objectName": "org.pih.warehouse.product.Product",
      "rejectedValue": null
    }
  ],
  "errorMessages": [
    "Property [category] of class [Product] cannot be null",
    "Property [Name] of class [Product] cannot be null"
  ]
}
```

Reply:
@DominikaHatala Can you revert your changes and see if the errorMessages gives you something close to what you need? If so then we can just edit the error message in messages.properties to conform to the requirements.
Which I think will be to change:
```
shipmentItem.quantity.cannotExceedAvailableQuantity=Quantity picked ({0}) cannot exceed available quantity ({1}) for product code ''{2}'' and lot number ''{3}'' at origin ''{4}'' bin ''{5}''
```
to something like
```
shipmentItem.quantity.cannotExceedAvailableQuantity=Shipping quantity ({0}) can not exceed on hand quantity ({1}) for product code ''{2}'' and lot number ''{3}'' at origin ''{4}'' bin ''{5}''. This can occur if changes were made to inventory after this shipment was picked but before it shipped. To move forward, please remove the lines above from the shipment or reduce to reflect current QOH.
```
Kelsey was expecting the API to send back all validation errors at once but it looks like we short circuit the loop and throw the validation exception after the first one. Since this was a 2-pointer let's give her the option of creating a new ticket to show all errors at once.